core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/ArraySortAfterArrayDistinct.java

```diff
@@ -0,0 +1,129 @@
+import io.trino.spi.type.Type;
+import io.trino.sql.planner.FunctionCallBuilder;
+import io.trino.sql.planner.TypeAnalyzer;
+import io.trino.sql.planner.iterative.Rule;
+import io.trino.sql.tree.Expression;
+import io.trino.sql.tree.ExpressionTreeRewriter;
+import io.trino.sql.tree.FunctionCall;
+import io.trino.sql.tree.NodeRef;
+import io.trino.sql.tree.QualifiedName;
+import io.trino.sql.tree.SymbolReference;
+import org.assertj.core.util.Lists;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import static com.google.common.base.Preconditions.checkState;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static java.util.Objects.requireNonNull;
+
+public class ArraySortAfterArrayDistinct
+        extends ExpressionRewriteRuleSet
+{
+    public ArraySortAfterArrayDistinct(Metadata metadata, TypeAnalyzer typeAnalyzer)
+    {
+        super(createRewrite(metadata, typeAnalyzer));
+    }
+
+    // This rewrite doesn't work on aggregation clauses
+    @Override
+    public Set<Rule<?>> rules()
+    {
+        return ImmutableSet.of(
+                projectExpressionRewrite(),
+                filterExpressionRewrite(),
+                joinExpressionRewrite(),
+                valuesExpressionRewrite());
+    }
+
+    private static ExpressionRewriter createRewrite(Metadata metadata, TypeAnalyzer typeAnalyzer)
+    {
+        requireNonNull(metadata, "metadata is null");
+        requireNonNull(typeAnalyzer, "typeAnalyzer is null");
+
+        return (expression, context) -> {
+            if (expression instanceof SymbolReference) {
+                return expression;
+            }
+            return ExpressionTreeRewriter.rewriteWith(new Visitor(context, typeAnalyzer, metadata), expression);
+        };
+    }
+
+    private static class Visitor
+            extends io.trino.sql.tree.ExpressionRewriter<Void>
+    {
+        private final Rule.Context context;
+        private final TypeAnalyzer typeAnalyzer;
+        private final Metadata metadata;
+
+        public Visitor(Rule.Context context, TypeAnalyzer typeAnalyzer, Metadata metadata)
+        {
+            this.context = context;
+            this.typeAnalyzer = typeAnalyzer;
+            this.metadata = metadata;
+        }
+
+        @Override
+        public Expression rewriteFunctionCall(FunctionCall node, Void context, ExpressionTreeRewriter<Void> treeRewriter)
+        {
+            if (isRequiredFunctionCall(node, ArrayDistinctFunction.NAME)
+                    && node.getArguments().size() > 0
+                    && node.getArguments().get(0) instanceof FunctionCall) {
+                FunctionCall arraySortFunctionCall = (FunctionCall) node.getArguments().get(0);
+                if (isRequiredFunctionCall(arraySortFunctionCall, ArraySortFunction.NAME)) {
+                    checkState(node.getArguments().size() == 1, "Function array_distinct called with unsupported number of arguments");
```

Comment:
After a second glance, I think that we should check 
`node.getArguments().size() == 1` and `arraySortArguments.size() <= 2` and just bail out if they don't hold, instead of throwing error. This is not responsibility of this rewrite to validate function calls.